File: AtCoder/ABC380/F_Exchange_Game.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, m, l;
    cin >> n >> m >> l;
    vector<int> cards(n + m + l);
    for (int i = 0; i < n + m + l; i++) cin >> cards[i];
    int mask = (1 << (n + m + l)) - 1;
    int a = 0, b = 0;
    for (int i = 0; i < n; i++) a |= (1 << i);
    for (int i = n; i < n + m; i++) b |= (1 << i);

    unordered_map<int, int> memo; // Memoization
    auto f = [&](auto &&f, int a, int b, bool flag) -> bool {
        int key = a << (n + m + l + 1) | b << 1 | flag;
        if (memo.count(key)) return memo[key];

        int cur = flag ? a : b; // 當前輪到某玩家出牌
        if (cur == 0) return false; // 當前玩家沒有牌可打就輸了
        int table = ~(a | b) & mask; // 桌面上的牌是兩個玩家手牌以外的牌
        
        int n_a, n_b, n_table;
        for (int i = 0; i < n + m + l; i++) { // 枚舉要打出的牌
            if (!((cur >> i) & 1)) continue; // 如果當前玩家沒有這張牌則跳過

            if (flag) {
                n_a = a & ~(1 << i);
                n_b = b;
            } else {
                n_a = a;
                n_b = b & ~(1 << i);
            }

            n_table = table | (1 << i); // 桌面上的牌加上剛剛打出的牌
            if (!f(f, n_a, n_b, flag ^ 1)) return memo[key] = true; // 不回收牌
            for (int j = 0; j < n + m + l; j++) {
                if (((n_table >> j) & 1) && (cards[j] < cards[i])) { // 這張牌在桌面上，且比剛剛打出的牌小，可以回收
                    if (flag) {
                        if (!f(f, n_a | (1 << j), n_b, flag ^ 1)) return memo[key] = true;
                    } else {
                        if (!f(f, n_a, n_b | (1 << j), flag ^ 1)) return memo[key] = true;
                    }
                }
            }
        }
        return memo[key] = false;
    };
    cout << (f(f, a, b, 1) ? "Takahashi" : "Aoki") << endl;

    return 0;
}

File: AtCoder/ABC380/F_Exchange_Game.py
"""
注意到 N + M + L <= 12，因此可以考慮狀壓DP，維護目前手牌以及輪到誰的回合三個狀態。

枚舉當前玩家手牌中可以出的卡，接著可以再選擇不回收任何牌、或再次枚舉牌桌上比打出牌更小的牌回收
如果這些可能中存在任何一種可以使對方贏不了，則自己可以贏；否則就換自己贏不了。
"""

import sys
from functools import cache
sys.setrecursionlimit(10**6)

N, M, L = map(int, input().split())
A = list(map(int, input().split()))
B = list(map(int, input().split()))
C = list(map(int, input().split()))

cards = A + B + C

mask = (1 << (N + M + L)) - 1
a = 0
for i in range(N):
    a |= (1 << i)
b = 0
for i in range(N, N + M):
    b |= (1 << i)

# f(a, b, flag) 表示當前輪到某玩家出牌，a 和 b 分別是兩個玩家的手牌，返回某玩家是否能贏
@cache
def f(a, b, flag):
    table = (~(a | b)) & mask # 桌面上的牌是兩個玩家手牌以外的牌

    cur = a if flag else b

    # 當前玩家沒有牌可打就輸了
    if cur == 0:
        return False

    # 枚舉要打出的牌
    for i, x in enumerate(cards):
        if (cur >> i) & 1 == 0:
            continue

        if flag:
            n_a = a & ~(1 << i)
            n_b = b
        else:
            n_a = a
            n_b = b & ~(1 << i)

        n_table = table | (1 << i) # 桌面上的牌加上剛剛打出的牌

        if not f(n_a, n_b, flag ^ 1): # 不回收牌
            return True

        for j, y in enumerate(cards):
            if ((n_table >> j) & 1) and (y < x): # 這張牌在桌面上，且比剛剛打出的牌小，可以回收
                if flag:
                    if not f(n_a | (1 << j), n_b, flag ^ 1):
                        return True
                else:
                    if not f(n_a, n_b | (1 << j), flag ^ 1):
                        return True
    return False

print("Takahashi" if f(a, b, 1) else "Aoki")